Add CRD API Documentation

ihcsim opened the pull request:
This PR adds the CRD API documentation to the doc site. It introduces a new script named `hack/gen-api.sh` that uses the [gen-crd-api-reference-docs](https://github.com/ahmetb/gen-crd-api-reference-docs/) tool to parse the CRD Go types found in the [external-snapshotter](https://github.com/kubernetes-csi/external-snapshotter/tree/master/client/apis/volumesnapshot/v1) repository, and writes the generated documentation to the `book/src/api.md` file.

Fix https://github.com/kubernetes-csi/external-snapshotter/issues/102.

Test output:

![image](https://user-images.githubusercontent.com/1330522/159410703-1acb2826-16b9-4945-8b32-ea8ea3b810eb.png)

xing-yang commented:
/assign

xing-yang commented:
/release-note-none

xing-yang commented on README.md:
````diff
@@ -8,6 +8,14 @@ To make changes to the documentation, modify the files in [book/src](https://git
+./hack/gen-api.sh
+```
+
+The script uses the [gen-crd-api-reference-docs](https://github.com/ahmetb/gen-crd-api-reference-docs/) tool to generate a markdown document for the [CSI CRD API](https://github.com/kubernetes-csi/external-snapshotter/tree/master/client/apis/volumesnapshot/v1). See the script for more information and supported configuration.
````
CSI CRD API -> VolumeSnapshot CRD API

xing-yang commented on book/src/SUMMARY.md:
```diff
@@ -46,4 +46,6 @@
     - [Unit Testing](unit-testing.md)
     - [Functional Testing](functional-testing.md)
 - [Drivers](drivers.md)
+- [Reference]()
```
This should not be empty in ()?

ihcsim replied:
mdbook currently doesn't allow non-link summary item. See https://github.com/rust-lang/mdBook/issues/483. I can add a new reference.md, if desired.

xing-yang commented on book/src/api.md:
```diff
@@ -0,0 +1,1055 @@
+# API Documentation
+
```
Can you add a sub-heading here for VolumeSnapshot?  We can add other CRDs later.

ihcsim replied:
The tool already generates subheadings for all the included CRDs like VolumeSnapshot, VolumeSnaphotContent and VolumeSnapshotClass:

![image](https://user-images.githubusercontent.com/1330522/163304478-72aac1fb-e793-47c6-9ab7-48966ca7da65.png)

![image](https://user-images.githubusercontent.com/1330522/163304547-0da80e59-ae63-4ce5-b654-db0fa9aaca9d.png)

![image](https://user-images.githubusercontent.com/1330522/163304573-57289bd7-ae4f-4782-83e5-133cea703af6.png)

To confirm my understanding, you would like me to add specifically VolumeSnapshot as a 2nd level subheading?

ihcsim replied:
I think I see what you mean. There are other non-snapshot CRDs in CSI. Yes, I can group all these snapshot CRDs under VolumeSnapshot.

xing-yang commented on hack/gen-api.sh:
```diff
@@ -0,0 +1,43 @@
+  -out-file="out.html"
+
+# add title to page
+printf "# API Documentation\n\n%s" "$(cat out.html)" > "$curr_dir"/book/src/api.md
```
Maybe have "api" as a folder and add "volumesnapshot.md" for snapshot APIs.  This way it will be easy to add another doc for other CRDs later.

ihcsim replied:
The current api.md contains other snapshot CRDs like VolumeSnapshotContent and VolumeSnapshotClass too. AIUI, the doc generation tool doesn't support splitting up multiple CRD `struct` in the same `types.go` file into multiple files.

xing-yang commented:
/assign @jingxu97

ihcsim commented:
@xing-yang I added a new section called "API Reference", with "Volume Snapshot" as its subsection. This should allow us to extend the list in the future to include other CRDs/APIs. All the `.md` files are also moved to the `src/api` folder. Let me know if this matches what you have in mind.

![image](https://user-images.githubusercontent.com/1330522/163306989-6702eaf2-81f4-43b2-a766-4873a4c98407.png)

xing-yang commented on book/src/api/volume-snapshot.md:
```diff
@@ -0,0 +1,1036 @@
+Exactly one of its members must be set.
+Members in VolumeSnapshotContentSource are immutable.
+TODO(xiangqian): Add a webhook to ensure that VolumeSnapshotContentSource members
+will be immutable once specified.</p>
```
Can you submit a PR in external-snapshotter to get rid of this TODO message?  It is no longer needed.
https://github.com/kubernetes-csi/external-snapshotter/blob/master/client/apis/volumesnapshot/v1/types.go#L344-L345

After that, re-run the script to generate the doc.

ihcsim replied:
I opened the PR at https://github.com/kubernetes-csi/external-snapshotter/pull/697.

ihcsim replied:
@xing-yang I have re-generated the doc to include the changes in https://github.com/kubernetes-csi/external-snapshotter/commit/b20011c84b1128ff21f432f3a448b63cc4b69f5b. PTAL.

xing-yang commented:
/lgtm
/approve
